### PR TITLE
S3 asf fixes detected by Terraform test suite

### PR DESCRIPTION
**Pull request by steffyP**
This PR contains several fixes for S3 ASF that were detected by the terraform test suite:

- Website Hosting: fix handling for empty `ReplaceKeyPrefixWith` for redirect-routing rule (in that case it should remove the the `KeyPrefixEquals` from the original url and then do the redirect)
- Bucket name with slashes: fix parsing issue; preserve name for keys that end with slash
- Patch to change casing of `content-lanugage` , `content-disposition` and `cache-control` to be picked up by parser
- Patch for bucket-key-enbled header `x-amz-server-side-encryption-bucket-key-enabled` which is returned as camel-case by moto, but should be lowercase to be picked up by the parser
- fix resetting of `content-length` for HEAD requests that are forwarded using the `Proxy` for vhost-style requests.
- fix issues with `delete-objects`:
   - query parameter was added twice to the url for vhost-style forwarding
   - add moto-patch to include old fix from legacy s3 provider
   - remove trailing `=` from the path (`/<bucket>/?delete=` but expects `/<bucket>/?delete`)
   - add notifications for `delete-objects`
   - implement `quiet` mode for `delete-objects`
- fix bucket replication and added tests
- add spec-patch for `log-delivery-write`, which was missing from the `BucketCannedACL`

## TODO before merge:

- [x]  remove last commit, which excludes some cors-related tests, that were timing out (CORS not yet implemented)
- [x] rebase with master (remove ASF enabled by default)

### CODE REVIEW

File: localstack/services/s3/utils.py
Comment (bentsku) [inline localstack/services/s3/utils.py:52]: We could add it with a patch I guess, weird it's not part of it.

File: localstack/aws/protocol/parser.py
Comment (thrau) [inline localstack/aws/protocol/parser.py:257]: predictably i'd prefer encapsulating this in the s3 parser somehow :D i don't have a good suggestion at the moment, will take a look after the review, maybe @alexrashed also has some ideas

File: localstack/http/client.py
Comment (thrau) [inline localstack/http/client.py:54]: good catch!

**Review by thrau — APPROVED** [review]
LGTM! i don't have a good suggestion on how to get the s3 hack out of the parser right now, so i'd be OK with merging it for now with the FIXME. i'll let alex make the final call though.

File: localstack/aws/protocol/parser.py
Comment (alexrashed) [reply localstack/aws/protocol/parser.py:257]: To be host, the `service_name` is something we desperately tried to avoid in these "generalized" parsers.
Since there is already an `S3RequestParser`, we could check this case and modifying the `uri_params` and afterwards calling the super method. Something like this:
```python
class S3RequestParser(RestXMLRequestParser):
    def _parse_shape(
        self, request: HttpRequest, shape: Shape, node: Any, uri_params: Mapping[str, Any] = None
    ) -> Any:
        if shape is not None and uri_params is not None \
                and shape.serialization.get("location") == "uri" \
                and shape.serialization.get("name") == "Key" \
                and request.url.split("?")[0].endswith(f"{uri_params['Key']}/"):
            uri_params = dict(uri_params)
            uri_params["Key"] = uri_params["Key"] + "/"
        return super()._parse_shape(request, shape, node, uri_params)
```

Another option would be to externalize the location parsing to a separate function, but this doesn't seem to be worth the effort / complexity.

I'm happy to assist if I can be of service!

**Review by alexrashed — APPROVED** [review]
Wow! This is great! The S3 ASF migration really goes to the limits (and beyond) of ASF. 🚀 
It would be nice if we could externalize the S3 specific edge case in the parser, but this also shouldn't be a blocker (I could also take care of that another time).

**Comment by steffyP** [comment]
Moved the specific s3-parsing (for the special case with trailing "/" in the key name) to the S3RequestParser as suggested by @alexrashed (thank you! 🙂) 

Removing the commit with the excluded CORS tests and rebasing with master now (to check that the tests still work with the old s3 provider), before merging.

**Pull request merged**
